documentation/specs/multithreaded-msbuild.md

```diff
@@ -0,0 +1,268 @@
+# Multithreaded MSBuild
+
+Currently, MSBuild supports parallel builds (a critical feature for a build system) by spawning worker processes. This made adoption easier because it didn't impose any requirements on tasks: they continue to own the whole process while they are executing. But it's a pretty strange design decision in the modern age, where we assume things are multithreaded and async.
+
+## Current state
+
+There are several components of MSBuild that are relevant to parallel builds:
```

Comment:
can we mention existing synchronization mechanisms when tasks attempt to access the same resource simultaneously?

Reply:
@AR-May mentioned retry mechanisms on the simultaneous access to the shared resources. Did I got you right, Alina?

Reply:
We were talking about example "Copy task accessing the file system", if I remember correctly. But in handling simultaneous access to some file there is no difference between how we handle it with multiple processes or multiple threads - we just have retry logic for opening files.

Reply:
But for different types of resources, we might behave differently, retry logic is not for any case. For some types of common resources approach is changed - like access to the environment variables and for some it did not.